payjoin/src/uri.rs

```diff
@@ -1,105 +1,173 @@
+use bitcoin::address::{Error, NetworkChecked, NetworkUnchecked, NetworkValidation};
+use bitcoin::{Address, Amount, Network};
 use url::Url;
 
-pub enum Payjoin {
-    Supported(PayjoinParams),
-    V2Only(PayjoinParams),
-    Unsupported,
+/// Payjoin Uri represents a bip21 uri with additional
+/// payjoin parameters.
+pub struct PayjoinUri<
+    'a,
+    N: NetworkValidation,
+    P: DeserializeParams<'a> + DeserializationError + Sized,
+> {
+    inner: bip21::Uri<'a, N, P>,
 }
 
-impl Payjoin {
-    pub fn pj_is_supported(&self) -> bool {
-        match self {
-            Payjoin::Supported(_) => true,
-            Payjoin::V2Only(_) => true,
-            Payjoin::Unsupported => false,
-        }
+impl<'a, N: NetworkValidation, P: DeserializeParams<'a> + DeserializationError + Sized>
+    From<bip21::Uri<'a, N, P>> for PayjoinUri<'a, N, P>
+{
+    fn from(value: bip21::Uri<'a, N, P>) -> Self { Self { inner: value } }
+}
+
+impl<'a> PayjoinUri<'a, NetworkUnchecked, Payjoin> {
+    /// Marks network of this address as checked.
+    pub fn assume_checked(self) -> PayjoinUri<'a, NetworkChecked, Payjoin> {
+        let checked_address = self.inner.address.assume_checked();
+        let mut uri = bip21::Uri::with_extras(checked_address, self.inner.extras);
+        uri.amount = self.inner.amount;
+        uri.label = self.inner.label;
+        uri.message = self.inner.message;
+        uri.into()
+    }
+    /// Checks whether network of this address is as required.
+    pub fn require_network(
+        self,
+        network: Network,
+    ) -> Result<PayjoinUri<'a, NetworkChecked, Payjoin>, Error> {
+        let checked_address = self.inner.address.require_network(network)?;
+        let mut uri = bip21::Uri::with_extras(checked_address, self.inner.extras);
+        uri.amount = self.inner.amount;
+        uri.label = self.inner.label;
+        uri.message = self.inner.message;
+        Ok(uri.into())
     }
 }
 
-pub struct PayjoinParams {
-    pub(crate) _endpoint: Url,
-    pub(crate) disable_output_substitution: bool,
+impl<'a, N: NetworkValidation> PayjoinUri<'a, N, Payjoin> {
+    /// Returns the receiver bitcoin address from this uri.
+    pub fn address(&self) -> &Address<N> { &self.inner.address }
+    /// Returns the amount in bitcoin from this uri.
+    pub fn amount(&self) -> Option<Amount> { self.inner.amount }
+    /// Returns the endpoint url from this uri.
+    pub fn endpoint(&self) -> Option<&Url> { self.inner.extras.endpoint() }
+    /// Sets the amount in bitcoin for this uri.
+    pub fn set_amount(&mut self, amount: &Option<Amount>) { self.inner.amount = *amount; }
+    /// Sets the label for this uri.
+    pub fn set_label(&mut self, label: String) {
+        self.inner.label = Some(bip21::Param::from(label));
+    }
+    /// Sets the message for this uri.
+    pub fn set_message(&mut self, message: String) {
+        self.inner.message = Some(bip21::Param::from(message));
```

Comment:
In Rust, the convention for naming setters is generally to not prefix them with set. I'm also not sure we need to implement these since what we had before e.g. `self.uri.extras.disable_output_substitution` was not broken, was it?

```rs
struct Example {
    value: i32,
}

impl Example {
    // Setter for `value`
    fn value(&mut self, value: i32) {
        self.value = value;
    }

    // Getter for `value`
    fn value(&self) -> i32 {
        self.value
    }
}
```